core/drivers/bcm_gpio.c

```diff
@@ -0,0 +1,148 @@
+
+	io_setbits32((vaddr_t)gc->base + offset, BIT(shift));
+}
+
+static inline void iproc_clr_bit(struct gpio_chip *gc, unsigned int reg,
+				 unsigned int gpio)
+{
+	unsigned int offset = IPROC_GPIO_REG(gpio, reg);
+	unsigned int shift = IPROC_GPIO_SHIFT(gpio);
+
+	io_clrbits32((vaddr_t)gc->base + offset, BIT(shift));
+}
+
+static void iproc_gpio_set(struct gpio_chip *gc, unsigned int gpio,
+			   enum gpio_level val)
+{
+	if (val == GPIO_LEVEL_HIGH)
+		iproc_set_bit(gc, IPROC_GPIO_DATA_OUT_OFFSET, gpio);
+	else
+		iproc_clr_bit(gc, IPROC_GPIO_DATA_OUT_OFFSET, gpio);
+}
+
+static enum gpio_level iproc_gpio_get(struct gpio_chip *gc, unsigned int gpio)
+{
+	unsigned int offset = IPROC_GPIO_REG(gpio, IPROC_GPIO_DATA_IN_OFFSET);
+	unsigned int shift = IPROC_GPIO_SHIFT(gpio);
+	int val;
+
+	val = io_read32((vaddr_t)gc->base + offset) & BIT(shift);
+
+	return val ? GPIO_LEVEL_HIGH : GPIO_LEVEL_LOW;
+}
+
+static void iproc_gpio_set_dir(struct gpio_chip *gc, unsigned int gpio,
+			      enum gpio_dir dir)
+{
+	if (dir == GPIO_DIR_OUT)
+		iproc_set_bit(gc, IPROC_GPIO_OUT_EN_OFFSET, gpio);
+	else
+		iproc_clr_bit(gc, IPROC_GPIO_OUT_EN_OFFSET, gpio);
+
+}
+
+static enum gpio_dir iproc_gpio_get_dir(struct gpio_chip *gc,
+					unsigned int gpio)
+{
+	unsigned int offset = IPROC_GPIO_REG(gpio, IPROC_GPIO_OUT_EN_OFFSET);
+	unsigned int shift = IPROC_GPIO_SHIFT(gpio);
+	uint32_t val;
+
+	val = io_read32((vaddr_t)gc->base + offset) & BIT(shift);
+
+	return val ? GPIO_DIR_OUT : GPIO_DIR_IN;
+}
+
+static enum gpio_interrupt iproc_gpio_get_itr(struct gpio_chip *gc,
+					  unsigned int gpio)
+{
+	unsigned int offset = IPROC_GPIO_REG(gpio, IPROC_GPIO_INT_MSK_OFFSET);
+	unsigned int shift = IPROC_GPIO_SHIFT(gpio);
+	uint32_t val;
+
+	val = io_read32((vaddr_t)gc->base + offset) & BIT(shift);
+
+	return val ? GPIO_INTERRUPT_ENABLE : GPIO_INTERRUPT_DISABLE;
+}
+
+static void iproc_gpio_set_itr(struct gpio_chip *gc, unsigned int gpio,
+			  enum gpio_interrupt ena_dis)
+{
+	if (ena_dis == GPIO_INTERRUPT_ENABLE)
+		iproc_set_bit(gc, IPROC_GPIO_OUT_EN_OFFSET, gpio);
+	else
+		iproc_clr_bit(gc, IPROC_GPIO_OUT_EN_OFFSET, gpio);
+
+}
+
+static const struct gpio_ops bcm_gpio_ops = {
+	.get_direction = iproc_gpio_get_dir,
+	.set_direction = iproc_gpio_set_dir,
+	.get_value = iproc_gpio_get,
+	.set_value = iproc_gpio_set,
+	.get_interrupt = iproc_gpio_get_itr,
+	.set_interrupt = iproc_gpio_set_itr,
+};
+KEEP_PAGER(bcm_gpio_ops);
+
+static void iproc_gpio_init(struct gpio_chip *gc, unsigned int paddr,
+			    unsigned int gpio_base, unsigned int ngpios)
+{
+	gc->base = (vaddr_t)phys_to_virt(paddr, MEM_AREA_IO_SEC);
+	gc->ops = &bcm_gpio_ops;
+	gc->gpio_base = gpio_base;
+	gc->ngpios = ngpios;
+
+	gpio_add_chip(gc);
+
+	EMSG("gpio probe success");
+}
+
+static TEE_Result bcm_gpio_init(void)
+{
+	struct gpio_chip *gc = malloc(sizeof(*gc));
+
+	if (gc == NULL)
+		return -1;
+
+	iproc_gpio_init(gc, SECURE_GPIO_BASE0, GPIO_NUM_START0, NUM_GPIOS0);
```

Comment:
Will this fit all BCM platforms? Maybe the plat-xxx/main.c (or like) could call bcm_gpio_init() providing **SECURE_GPIO_BASE0**/**...çSTART0**/**..._NUM**. This would prevent this driver from relying on **platform_config.h** and would help this driver to match several bcm platforms (if there are several).
This is only a suggestion. Maybe you prefer this driver to be specific to your currently existing platform(s).

Reply:
I thought of doing so. ie. call bcm_gpio_init() providing all details for required gpio chips from main.c   Later thought if I plan to derive these info from dt then I might want to keep it here it self.  Just to keep them all uniform driver_init() does the initialization.  
 The only challenge I have is ordering the driver_initxxx.  which I will have to use driver_init_late() for the consumer drivers. But if I soon face a road block then will move all inits to main.c and order them manually. Currently in spi_test spi driver init is done by PTA , so such use case is fine.

Reply:
These defines are included in #2927. If this gets merged before #2927, then this will also become a 'broken' commit.